### ISSUE TITLE
[Leo] Summarize selected text from context menu

### ISSUE THREAD

**Issue by yrliou** [brave/brave-browser#35640]
Add a new Leo tools submenu with a "Summarize" item in context menu when user selects some text in the page, which will open Leo do handle the request.
A setting should be added to enable/disable showing Leo tools in the context menu.

Here're all scenarios when context menu item is clicked and whether the whole page content will be sent.
1. Leo side panel was closed and user never send a request via side panel before in this page (no conversation history), no page content will be sent.
2. Leo side panel was closed but user has sent a request in this page before with page context linked, we keep it as is so page content will be sent again. Note that it also means the same page context was already sent in a previous request.
3. Leo side panel was closed but user has sent a request in this page before without page context linked, no page content will be sent.
4. Leo side panel is already open, page content is currently linked, page content will be sent.
5. Leo side panel is already open, page content is currently not linked, no page content will be sent.
In other words, if you starts with context menu, no page context will be sent. If you starts with a side panel, then we follow the status of the side panel, since this will be inserted as a linear message into the current conversation in the side panel.

**Comment by MadhaviSeelam** [comment]
Verification `In progress` using
```
Brave | 1.64.86 Chromium: 122.0.6261.69 (Official Build) beta (64-bit)
-- | --
Revision | 865fcaaa1d24fb4d6eb5372332ff8687f75b8c59
OS | Windows 11 Version 23H2 (Build 22631.3155)
```
Case 1:  Default option to hide Leo tools in the context menu
<details>

1. Installed 1.64.86
2. launched Brave
3. opened `brave://settings/leo-assistant` in a new tab
4. confirmed `Show Leo in the context menu on websites` toggle is `on`
5. loaded an article in a new tab https://www.cnn.com/2024/02/26/world/japan-moon-sniper-slim-wakes-up-images-scn/index.html
6. selected text and right clicked to open context menu
7. confirmed `Leo tools` submenu with a "Summarize" item in context menu
8. clicked `Summarize`
8. Leo panel opened with privacy agreement
9. clicked `I understand`
10. verified the selected text is summarized in the Leo panel
11. switched to Claude Instant by Anthropic` from 3 dots Leo setting menu
12. highlighted text again and opened context menu
13. verified the selected text is summarized in the Leo panel

#### Confirmed selected text summarized successfully

step 4 | step 6-7 | step 8-9 | step 10 | step 11-12 | step 13
---------| -------- | --------- | -------- | ----------| -------
<img width="955" alt="image" src="https://github.com/brave/brave-browser/assets/98358127/ba3ed0a7-ea10-4bc7-baf3-72a67df18ab7">|<img width="956" alt="image" src="https://github.com/brave/brave-browser/assets/98358127/4e95f787-e073-42c1-83af-a473a6a851a3">|<img width="959" alt="image" src="https://github.com/brave/brave-browser/assets/98358127/791d1eca-0599-4c97-95b9-18746a2efa77">|<img width="956" alt="image" src="https://github.com/brave/brave-browser/assets/98358127/92ad4cd4-e9af-4ee7-ac1c-489d0af85498">|<img width="956" alt="image" src="https://github.com/brave/brave-browser/assets/98358127/4b8eb7cc-5567-499c-aa68-a3eab45313bc">|<img width="956" alt="image" src="https://github.com/brave/brave-browser/assets/98358127/0b464adb-6126-4be4-ad1a-4e6501b94201">
 </details>
 
 Case 2:  Disable the option to show Leo tools in the context menu
  <details>
  
1. new profile
2. launched Brave
3. opened `brave://settings/leo-assistant` in a new tab
4. disabled `Show Leo in the context menu on websites` toggle 
5. visited an article https://www.usatoday.com/story/travel/2024/02/26/tiktok-sober-gen-z-travel-trends/72684017007/
6. selected text and right clicked to open context menu

#### Confirmed `Leo tools` submenu with a "Summarize" item is not shown in the context menu

step 3 | step 6
---- | ---
<img width="956" alt="image" src="https://github.com/brave/brave-browser/assets/98358127/ccc605ee-925d-4b67-bf51-e0547072f2da">|<img width="956" alt="image" src="https://github.com/brave/brave-browser/assets/98358127/c9444aa7-c371-43d0-bfc7-8976901e124c">
 </details>
 
Case 3: Opt-in first before summarizing via Leo tools menu

 <details>
1. new profile
2. launched Brave
3. clicked `Leo` icon in the sidebar
4. entered `hi` in the prompt field
5. clicked `I understand` button
6. visited an article https://www.bbc.com/news/world-us-canada-68407453
7. highlighted text and right clicked open context menu
8. selected `Leo tools` menu >> `Summarize`

#### Confirmed selected text summarized successfully

<img width="953" alt="image" src="https://github.com/brave/brave-browser/assets/98358127/19226df3-58c4-40ca-a45a-b4ac20d75a0d>|<img width="953" alt="image" src="https://github.com/brave/brave-browser/assets/98358127/a0cb35b9-72ed-4eb4-af62-903ba5cdd90d">|<img width="953" alt="image" src="https://github.com/brave/brave-browser/assets/98358127/e59bed62-0386-4df5-bedb-5f3f550e6d74">
 </details>
 
 Case 3: Leo side panel was closed but user has sent a request in this page before with page context linked

1. new profile
2. launched Brave
3. clicked `Leo` icon in the sidebar
4. entered `hi` in the prompt field
5. opened an article https://www.usatoday.com/story/news/weather/2024/02/26/severe-weather-outlook-across-us/72747218007/
6. highlight some text to open context menu
7. clicked `I understand` button
8. scrolled down the page and selected some more text

#### Confirmed the excerpt summarized successfully
#### Confirmed Leo panel stayed open
#### Confirmed selected additional text was summarized in the same Leo panel

example | example | example 
----- | ----- | ---
![image](https://github.com/brave/brave-browser/assets/98358127/840c2ace-a4c7-46b5-a68b-206205124d4f)|![image](https://github.com/brave/brave-browser/assets/98358127/c184c0bd-0040-48a2-b2ed-fb1ad68c7c0d)|![image](https://github.com/brave/brave-browser/assets/98358127/e4a93fd9-8e38-488e-9067-a559abaad22a)|![image](https://github.com/brave/brave-browser/assets/98358127/3b7d1d95-e848-4378-883f-2dc89dc3578b)|![image](https://github.com/brave/brave-browser/assets/98358127/0117ec4f-aa39-4f1a-9bea-4992460bdb67)|![image](https://github.com/brave/brave-browser/assets/98358127/0f8049fd-60c8-4479-937e-26333105a431)
